lsp-types/src/Language/LSP/Types/Uri/OsPath.hs

```diff
@@ -0,0 +1,47 @@
+{-# LANGUAGE CPP                 #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+#if MIN_VERSION_filepath(1,4,100)
+#define OS_PATH 1
+#endif
+
+module Language.LSP.Types.Uri.OsPath
+  (
+#ifdef OS_PATH
+    osPathToNormalizedFilePath
+  , normalizedFilePathToOsPath
+#endif
+  ) where
+
+#ifdef OS_PATH
+
+import           Control.DeepSeq        (NFData, force)
+import           Control.Exception      hiding (try)
+import           Control.Monad.Catch
+import           Language.LSP.Types.Uri
+import           System.IO.Unsafe       (unsafePerformIO)
+import           System.OsPath
+
+{-|
+Constructs 'NormalizedFilePath' from 'OsPath'. Throws 'IOException' if the conversion fails.
+-}
+osPathToNormalizedFilePath :: MonadThrow m => OsPath -> m NormalizedFilePath
+osPathToNormalizedFilePath = fmap toNormalizedFilePath . unsafePerformIO' . decodeFS
```

Comment:
Instead of `decodeFS` you could use the pure `decodeWith` and isolate the `unsafePerformIO` in a single global:

```
osPathToNormalizedFilePath :: MonadThrow m => OsPath -> m NormalizedFilePath
 osPathToNormalizedFilePath = fmap toNormalizedFilePath . unsafePerformIO' . decodeWith systemEnc

{-# NOINLINE systemEnc #-}
systemEnc :: TextEncoding
systemEnc = unsafePerformIO getFileSystemEncoding
```

Reply:
Applied in #453. Thank you!